web/app/datasets/handle_signals.py

```diff
@@ -13,14 +13,105 @@
 - allow signal API to post a message to process
 - add a proper task queue (like Celery)
 """
+# Assumptions:
+# 1) Each signal will be sent to 1 API at most (underlies both table design and
+#    routing implementation.
+
+
 import os
+import datetime
+import logging
 
 import requests
+from django.conf import settings
+from django.utils import timezone
+
+from datasets.models import MessageLog
+from datasets.external.base import get_handler
+
+# -- setup logging --
+LOG_FORMAT = '%(asctime)-15s - %(name)s - %(message)s'
+logging.basicConfig(format=LOG_FORMAT, level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+# -- Datapunt internal Signalen in Amsterdam API endpoints --
+#    note: should be set by Ansible or docker-compose.yml
+SIGNALS_API_BASE = os.getenv('SIGNALS_API_BASE', 'https://acc.api.data.amsterdam.nl')
+
+
+def _get_session_with_retries():
+    """
+    Get a requests Session that will retry 5 times on a number of HTTP 500 statusses.
+    """
+    session = requests.Session()
+
+    retries = Retry(
+        total=5,
+        backoff_factor=0.1,
+        status_forcelist=[500, 502, 503, 504],
+        raise_on_status=True
+    )
+    adapter = HTTPAdapter(max_retries=retries)
+    session.mount('https://', adapter)
+    session.mount('http://', adapter)
+
+    return session
+
+
+def _batch_signals(signals):
+    """
+    Access the Signalen in Amsterdam API, retrieve signals.
+    """
+    next_page = SIGNALS_API_BASE + '/signal/'
+
+    with _get_session_with_retries() as session:
+        result = session.get(next_page)
+        yield result.json()['results']
+
+        next_page = results.json()['_links']['next']['href']
+        if next_page == None:
+            raise StopIteration
+
+
+def _call_external_apis(signals):
+    """
+    Call external APIs for each of the signals.
+
+    Note signals are expected as dictionaries, not objects.
+    """
+    for s in signals:
```

Comment:
for signal in signals reads a little better

Reply:
changed s to signal as requested